cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -192,191 +108,102 @@ struct dispatch_quadtree_point_in_polygon {
-    thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
-                           d_num_hits.begin(),
-                           d_num_hits.end(),
-                           d_num_hits.begin());
-
-    uint32_t total_hits = d_num_hits.back_element(stream);
-
-    auto poly_index_col  = make_fixed_width_column<uint32_t>(total_hits, stream, mr);
-    auto point_index_col = make_fixed_width_column<uint32_t>(total_hits, stream, mr);
-
-    // write output directly to poly_index_col and point_index_col columns
-    auto d_poly_index_col  = cudf::mutable_column_device_view::create(*poly_index_col, stream);
-    auto d_point_index_col = cudf::mutable_column_device_view::create(*point_index_col, stream);
-
-    quad_pip_phase2_kernel<T, threads_per_block>
-      <<<num_valid_pair, threads_per_block, 0, stream>>>(d_pq_poly_idx.data(),
-                                                         d_pq_quad_idx.data(),
-                                                         d_quad_offset.data(),
-                                                         d_quad_length.data(),
-                                                         *d_quad_offsets,
-                                                         *d_point_indices,
-                                                         *d_point_x,
-                                                         *d_point_y,
-                                                         *d_poly_offsets,
-                                                         *d_ring_offsets,
-                                                         *d_poly_points_x,
-                                                         *d_poly_points_y,
-                                                         d_num_hits.data(),
-                                                         *d_poly_index_col,
-                                                         *d_point_index_col);
+                        poly_and_point_indices,
+                        poly_and_point_indices + num_total_points,
+                        test_poly_point_intersection<T, decltype(point_xys_iter)>{
+                          point_xys_iter,
+                          *cudf::column_device_view::create(poly_offsets, stream),
+                          *cudf::column_device_view::create(ring_offsets, stream),
+                          *cudf::column_device_view::create(poly_points_x, stream),
+                          *cudf::column_device_view::create(poly_points_y, stream)}));
+
+    // Allocate output columns for the number of pairs that intersected
+    auto poly_idx_col  = make_fixed_width_column<uint32_t>(num_intersections, stream, mr);
+    auto point_idx_col = make_fixed_width_column<uint32_t>(num_intersections, stream, mr);
+
+    // Note: no need to resize `poly_idxs` or `point_idxs` if we set the end iterator to
+    // `idxs.begin() + num_intersections`.
+
+    // populate the polygon indices column
+    thrust::copy(rmm::exec_policy(stream)->on(stream),
```

Comment:
Do you need the extra copy? Could you use `copy_remove_if` above?

Reply:
Oh sweet, I didn't know about `copy_remove_if`! :tada:

Reply:
I can't use `copy_remove_if` because the size of the output columns is determined by the number of elements removed by `remove_if`.